### PR TITLE
Interpolation with coarsening for general staggering

### PR DESCRIPTION
**Pull request by EZoni**
**Note** This PR is related to PR #856 and issue #797. CPU/GPU regression tests ongoing.

**Goal**

Implement interpolation (namely, averaging over either 1 point or 2 equally distant points) with coarsening for general staggering of input and output data.

**Description**

New functions defined in the namespace `Average`: `CoarsenAndInterpolate`, `CoarsenAndInterpolateLoop` and `CoarsenAndInterpolateKernel`. The following cases are covered:
- fine nodal to coarse nodal
- fine centered to coarse nodal
- fine nodal to coarse centered
- fine centered to coarse centered

The labels above are meaningful in 1D: extending the same idea to 2D and 3D, face-centered data and edge-centered data are covered as well. Previous calls to `amrex::average_down` (should not give the exact same result) and `Average::ToCellCenter` (should give the same result) have been replaced with calls to `Average::CoarsenAndInterpolate`.

**Testing the PR**

The core loop implemented in `Average::CoarsenAndInterpolateKernel` has been tested by checking the indices of the input data used for interpolation with the following Python script:

```py
import numpy as np

# ic: index of the destination coarse array
# si: staggering of the source fine MultiFab
# so: staggering of the destination coarse MultiFab
# cr: coarsening ratio
def indices_onestag( ic, si, so, cr ):
    if ( cr==0 or ( cr!=1 and np.ceil(np.log2(cr))!=np.floor(np.log2(cr)) ) ):
        print( '\n ABORT: coarsening ratio must be an integer power of 2' )
        return
    elif (cr==1):
        n    = 1+abs(si-so) #2-(2*si*so-si-so+1)
        imin = ic-so*(1-si)
    else:
        n    = 2-si
        imin = ic*cr+cr//2*(1-so)-(1-si)
    if ( si==1 and so==1 ):
        print( '\n fine nodal to coarse nodal (coarsening ratio={})'.format( cr ) )
    elif ( si==0 and so==1 ):
        print( '\n fine centered to coarse nodal (coarsening ratio={})'.format( cr ) )
    elif ( si==1 and so==0 ):
        print( '\n fine nodal to coarse centered (coarsening ratio={})'.format( cr ) )
    elif ( si==0 and so==0 ):
        print( '\n fine centered to coarse centered (coarsening ratio={})'.format( cr ) )
    print( ' interpolate over n={} points'.format( n ), end='' )
    print( ' to get value at i_c={}:'.format( ic ), end='' )
    for iref in range( n ):
        print( ' i_f={} '.format( imin+iref ), end='' )
    print()

def indices_allstag( ic, cr ):
    indices_onestag( ic, si=1, so=1, cr=cr ) # fine nodal    to coarse nodal
    indices_onestag( ic, si=0, so=1, cr=cr ) # fine centered to coarse nodal
    indices_onestag( ic, si=1, so=0, cr=cr ) # fine nodal    to coarse centered
    indices_onestag( ic, si=0, so=0, cr=cr ) # fine centered to coarse centered
```

The following schematic illustrates the meaning of interpolation with coarsening for general input and output staggering in 1D (from top to bottom: coarse centered data, coarse nodal data, fine centered data, fine nodal data; the indices `i_c` and `i_f` refer to coarse and fine data, respectively):
CoarsenAndInterpolate1D

### CODE REVIEW

File: Source/Utils/Average.H [hunk 1]
```diff
@@ -89,6 +89,55 @@ namespace Average{
                         const int ngrow,
                         const int scomp=0,
                         const int ncomp=1 );
+
+    AMREX_GPU_HOST_DEVICE
+    AMREX_FORCE_INLINE
+    Real FineToCoarse ( Array4<Real const> const& mf_fp_arr,
```
Comment (MaxThevenet) [inline Source/Utils/Average.H]: The name `FineToCoarse` is misleading, because in the rest of the code it refers to interpolation from the fine patch to the coarse patch of the same MR level, while this function simply interpolates from 1 resolution to another. Could it be another name? `Coarsen` for instance.

File: Source/Utils/Average.H [hunk 2]
```diff
@@ -92,13 +92,13 @@ namespace Average{
...
     AMREX_GPU_HOST_DEVICE
     AMREX_FORCE_INLINE
-    Real FineToCoarse ( Array4<Real const> const& mf_fp_arr,
-                        const IntVect stag,
-                        const int i,
-                        const int j,
-                        const int k,
-                        const int comp,
-                        const IntVect ratio )
+    Real Coarsen ( Array4<Real const> const& mf_fp_arr,
```
Comment (ax3l) [inline Source/Utils/Average.H:79]: would this be a good location for an extended doxygen string for the method?

File: Source/Utils/Average.H [hunk 3]
```diff
@@ -133,11 +133,11 @@ namespace Average{
         return volfrac * c;
     };
 
-    void FineToCoarse ( MultiFab& mf_cp,
-                        const MultiFab& mf_fp,
-                        const int scomp,
-                        const int ncomp,
-                        const IntVect ratio );
+    void Coarsen ( MultiFab& mf_cp,
```
Comment (ax3l) [inline Source/Utils/Average.H:120]: would this (or above) be a good location for an extended doxygen string for the method?

Comment (EZoni) [reply Source/Utils/Average.H:120]: Yes, thanks. This PR is still very much in progress as we are in the process of deciding what we want to implement with @MaxThevenet and @RevathiJambunathan. Once the implementation is completed I will add a Doxygen documentation for the new method(s) in this header file.

Comment (ax3l) [reply Source/Utils/Average.H:127]: Sounds awesome, thx! :)

File: Source/Utils/Average.H [hunk 4]
```diff
@@ -9,24 +9,19 @@ namespace Average{
...
+     *                      the data to be averaged
+     * \param[in] j         index along y to access the Array4 \c mf_in_arr containing
+     *                      the data to be averaged
+     * \param[in] k         index along z to access the Array4 \c mf_in_arr containing
+     *                      the data to be averaged
+     * \param[in] comp      index along the fourth component of the Array4 \c mf_in_arr
+     *                      containing the data to be averaged
+     *
      * \return averaged field at cell (i,j,k)
      */
```
Comment (EZoni) [inline Source/Utils/Average.H:26]: Reason for cleaning up this Doxygen lines: the web links I added in an earlier PR made the comments non-readable in the code itself, hence I removed them. These changes are not strictly related to this PR, but most of this file is being modified anyways and I'm taking the opportunity to make the Doxygen documentation uniform among the functions defined in this whole new namespace.

File: Source/Utils/Average.H [hunk 5]
```diff
@@ -57,38 +52,145 @@ namespace Average{
...
+     *                       object, extracted from its MultiFab, containing the data
+     *                       to be averaged
      * \param[in]     ncomp  optional number of components to loop over for both the
-     *                       input and the output
-     *                       <a href="https://amrex-codes.github.io/amrex/doxygen/structamrex_1_1Array4.html">Array4</a>
-     *                       objects, extracted from their respective
-     *                       <a href="https://amrex-codes.github.io/amrex/doxygen/classamrex_1_1MultiFab.html">MultiFab</a>
+     *                       input and the output Array4 objects, extracted from their
+     *                       respective MultiFab
      */
```
Comment (EZoni) [inline Source/Utils/Average.H:69]: Reason for cleaning up this Doxygen lines: the web links I added in an earlier PR made the comments non-readable in the code itself, hence I removed them. These changes are not strictly related to this PR, but most of this file is being modified anyways and I'm taking the opportunity to make the Doxygen documentation uniform among the functions defined in this whole new namespace.

File: Source/Utils/Average.cpp [hunk 6]
```diff
@@ -27,3 +27,99 @@ Average::ToCellCenter ( MultiFab& mf_out,
...
+    else
+    // Copy from component scomp of the fine FArrayBox into component 0 of the coarse
+    // FArrayBox because the coarse FArrayBox is a temporary FArrayBox starting at
+    // component 0 and is not part of the actual coarse MultiFab mf_cp
+    {
+        MultiFab coarsened_mf_fp( coarsened_mf_fp_ba, mf_fp.DistributionMap(), ncomp, 0, MFInfo(), FArrayBoxFactory() );
+        Average::CoarsenAndInterpolateLoop( coarsened_mf_fp, mf_fp, 0, scomp, ncomp, ratio );
+        // Copy starts at component 0 in the temporary source MultiFab coarsend_mf_fp
+        mf_cp.copy( coarsened_mf_fp, 0, dcomp, ncomp );
+    }
```
Comment (MaxThevenet) [inline Source/Utils/Average.cpp:124]: That's neat. Minor suggestion, just about comments:
```suggestion
    {
        // Cannot coarsen directly into a MultiFab with different BoxArray/DistributionMapping.
        // Hence, do it in two steps:

        // Temp MultiFab on coarsened version of mf_fp.boxArray(), with same distribution mapping.
        MultiFab coarsened_mf_fp( coarsened_mf_fp_ba, mf_fp.DistributionMap(), ncomp, 0, MFInfo(), FArrayBoxFactory() );
        // First, do the interpolation from mf_fp to coarsened_mf_fp (start writing into component 0).
        Average::CoarsenAndInterpolateLoop( coarsened_mf_fp, mf_fp, 0, scomp, ncomp, ratio );
        // Second, copy from coarsened_mf_fp to mf_cp (with different BoxArray/DistributionMapping).
        mf_cp.copy( coarsened_mf_fp, 0, dcomp, ncomp );
    }
```

File: Source/Utils/Average.cpp [hunk 7]
```diff
@@ -27,3 +27,99 @@ Average::ToCellCenter ( MultiFab& mf_out,
...
+    sc[2] = stag_cp[2];
+#endif
+
+    cr[0] = ratio[0];
+    cr[1] = ratio[1];
+#if   (AMREX_SPACEDIM == 2)
+    cr[2] = 1;
+#elif (AMREX_SPACEDIM == 3)
+    cr[2] = ratio[2];
+#endif
```
Comment (MaxThevenet) [inline Source/Utils/Average.cpp:71]: What about this? Trying to shorten and clarify.
```suggestion
    // Staggerings of input fine MultiFab and output coarse MultiFab
    IntVect sf = mf_fp.boxArray().ixType().ixType();
    IntVect sc = mf_cp.boxArray().ixType().ixType();
    IntVect cr = crse_ratio;
#if (AMREX_SPACEDIM == 2)
    sf[2] = 0;
    sc[2] = 0;
    cr[2] = 1;
#endif

    AMREX_ALWAYS_ASSERT_WITH_MESSAGE( mf_fp.nGrowVect() >= IntVect(stag_cp-stag_fp),
        "input MultiFab does not have enough guard cells for this interpolation" );
```

File: Source/Utils/Average.cpp [hunk 8]
```diff
@@ -27,3 +27,99 @@ Average::ToCellCenter ( MultiFab& mf_out,
...
     }
 }
+
+void
+Average::CoarsenAndInterpolateLoop ( MultiFab& mf_cp,
+                                     const MultiFab& mf_fp,
+                                     const int dcomp,
+                                     const int scomp,
+                                     const int ncomp,
+                                     const IntVect ratio )
```
Comment (MaxThevenet) [inline Source/Utils/Average.cpp:37]: I think it would be nice to use the name `crse_ratio`, as we use it in other places and it is more explicit.
```suggestion
                                     const IntVect& crse_ratio )
```

File: Source/Utils/Average.H [hunk 9]
```diff
@@ -57,38 +52,145 @@ namespace Average{
...
+    Real CoarsenAndInterpolateKernel ( Array4<Real const> const& mf_fp_arr,
+                                       const int sf[3],
+                                       const int sc[3],
+                                       const int cr[3],
+                                       const int i,
+                                       const int j,
+                                       const int k,
+                                       const int comp )
+    {
+        // Indices of coarse array
```
Comment (MaxThevenet) [inline Source/Utils/Average.H:105]:
```suggestion
        using namespace amrex;

        // Indices of coarse array
```

File: Source/Utils/Average.H [hunk 10]
```diff
@@ -57,38 +52,145 @@ namespace Average{
...
+        // Auxiliary integer variables
+        const int numx = num_pts[0];
+        const int numy = num_pts[1];
+        const int numz = num_pts[2];
+        const int imin = idx_min[0];
+        const int jmin = idx_min[1];
+        const int kmin = idx_min[2];
+
+        // Interpolate over points computed above
+        Real c = 0.0;
```
Comment (MaxThevenet) [inline Source/Utils/Average.H:134]:
```suggestion
        Real c = 0.0_rt;
```

File: Source/Utils/Average.H [hunk 11]
```diff
@@ -57,38 +52,145 @@ namespace Average{
...
+                                       const int j,
+                                       const int k,
+                                       const int comp )
+    {
+        // Indices of coarse array
+        const int ic[3] = { i, j, k };
+
+        // Auxiliary integer arrays
+        int num_pts[3];
+        int idx_min[3];
```
Comment (MaxThevenet) [inline Source/Utils/Average.H:110]:
```suggestion
        // Number of points for the interpolation, along each direction. 1 or 2.
        int num_pts[3];
        // Index of first point of src from which interpolation is done.
        int idx_min[3];
```

File: Source/Utils/Average.H [hunk 12]
```diff
@@ -57,38 +52,145 @@ namespace Average{
...
+     * \param[in] j         index along y of the coarsened Array4 to be filled
+     * \param[in] k         index along z of the coarsened Array4 to be filled
+     * \param[in] comp      index along the fourth component of the Array4 \c mf_fp_arr
+     *                      containing the data to be interpolated
+     *
+     * \return interpolated field at cell (i,j,k) of a coarsened Array4
+     */
+    AMREX_GPU_HOST_DEVICE
+    AMREX_FORCE_INLINE
+    Real CoarsenAndInterpolateKernel ( Array4<Real const> const& mf_fp_arr,
```
Comment (MaxThevenet) [inline Source/Utils/Average.H:96]: This is not a MultiFab, and maybe including `src` in the name would help, e.g.,
```suggestion
    Real CoarsenAndInterpolateKernel ( Array4<Real const> const& src_arr,
```

File: Source/Utils/Average.cpp [hunk 13]
```diff
@@ -27,3 +27,99 @@ Average::ToCellCenter ( MultiFab& mf_out,
...
+                                 const IntVect ratio )
+{
+    BL_PROFILE( "Average::CoarsenAndInterpolate" );
+
+    AMREX_D_TERM( AMREX_ALWAYS_ASSERT_WITH_MESSAGE( mf_fp.boxArray().coarsenable( ratio[0] ),
+                      "input MultiFab is not coarsenable along x" );,
+                  AMREX_ALWAYS_ASSERT_WITH_MESSAGE( mf_fp.boxArray().coarsenable( ratio[1] ),
+                      "input MultiFab is not coarsenable along y" );,
+                  AMREX_ALWAYS_ASSERT_WITH_MESSAGE( mf_fp.boxArray().coarsenable( ratio[2] ),
+                      "input MultiFab is not coarsenable along z" ); );
```
[the remark this answers is not in the document]
Comment (EZoni) [reply Source/Utils/Average.cpp:107]: Yes, thank you for the suggestion. There is indeed an overload of the method `coarsenable` that takes an `IntVect` as argument.

File: Source/Utils/Average.cpp (hunk 7 above)
Comment (EZoni) [reply Source/Utils/Average.cpp:71]: I'm afraid that this cannot be done. The reason why I used C++ arrays instead of objects of type `IntVect` is that `IntVect` is a dimension-dependent class representing an integer vector in `AMREX_SPACEDIM`-dimensional space. Hence, when `AMREX_SPACEDIM=2` assignments like `sf[2]=0;` would try to access invalid memory locations (and be aborted by the ASSERT call in `amrex/Src/Base/AMReX_IntVect.H:186`). The reason for having this logic before calling the kernel that performs the interpolation was to write only one kernel algorithm in 3D that does not distinguish explicitly between 2D and 3D. Does it make sense? I will try to think about other ways to shorten and clarify this piece of code.

File: Source/Utils/Average.H (hunk 9 above)
Comment (EZoni) [reply Source/Utils/Average.H:105]: This `using` declaration is already present at the beginning of the namespace `Average` in `Source/Utils/Average.H:8`. Is it OK to leave it there or should it be added within the scope of each function separately (when needed)?

File: Source/Utils/Average.cpp [hunk 14]
```diff
@@ -27,3 +27,99 @@ Average::ToCellCenter ( MultiFab& mf_out,
...
+void
+Average::CoarsenAndInterpolateLoop ( MultiFab& mf_cp,
+                                     const MultiFab& mf_fp,
+                                     const int dcomp,
+                                     const int scomp,
+                                     const int ncomp,
+                                     const IntVect ratio )
+{
+    // Staggerings of input fine MultiFab and output coarse MultiFab
+    const IntVect stag_fp = mf_fp.boxArray().ixType().ixType();
```
Comment (RevathiJambunathan) [inline Source/Utils/Average.cpp:40]:
```suggestion
    const IntVect stag_fp = mf_fp.boxArray().ixType().toIntVect();
    const IntVect stag_cp = mf_cp.boxArray().ixType().toIntVect();
```

File: Source/Utils/Average.H (hunk 12 above)
Comment (EZoni) [reply Source/Utils/Average.H:96]: Yes, I can do that. The reason for calling this argument `mf_fp_arr` (and similar variables in `Source/Utils/Average.cpp`) was to make clear by the name itself what is the `MultiFab` (`mf_fp` in this case) whose data (box-wise) are being pointed to by that `Array4`. I would try anyways to be consistent, hence choose something like `mf_src`, `arr_src` together with `mf_dst` and `arr_dst`.

File: Source/Utils/Average.cpp (hunk 14 above)
Comment (EZoni) [reply Source/Utils/Average.cpp:40]: Thanks for this suggestion, I'll change this to be compliant with the new convention adopted with #895.

File: Source/Utils/Average.H [hunk 15]
```diff
@@ -57,38 +52,146 @@ namespace Average{
...
+     *                           Array4 extracted from the coarsened MultiFab \c mf_cp
+     * \param[in]     crse_ratio coarsening ratio between the fine MultiFab \c mf_fp
+     *                           and the coarsened MultiFab \c mf_cp
+     */
+    void CoarsenAndInterpolateLoop ( MultiFab& mf_cp,
+                                     const MultiFab& mf_fp,
+                                     const int dcomp,
+                                     const int scomp,
+                                     const int ncomp,
+                                     const IntVect crse_ratio );
```
Comment (RevathiJambunathan) [inline Source/Utils/Average.H:169]: Can you provide a default a value for IntVect crse_ratio as amrex::IntVect(AMREX_D_DECL(1,1,1)). For cases, where, interpolation is needed without coarsening.

Comment (EZoni) [reply Source/Utils/Average.H:169]: Sure, I will do that. `amrex::IntVect(1)` should be enough for this purpose (see `amrex/Src/Base/AMReX_IntVect.H:100`).

Comment (RevathiJambunathan) [reply Source/Utils/Average.H:169]: Oh! that is neater! 👍

**Comment by EZoni** [comment]
> Looks great, thanks! See suggestions below. One point, though: the way you handle different BoxArrays is
> 
> ```
> CoarsenAndInterpolate{
>     if (same BA):
>         CoarsenAndInterpolateLoop
>     else:
>         create tmp
>         CoarsenAndInterpolateLoop
>         copy from tmp to dst
> }
> ```
> 
> It is fine, but it creates the weird function `CoarsenAndInterpolateLoop` with the exact same prototype but (somehow artificially) a different name. Should we consider another way, like
> 
> ```
> CoarsenAndInterpolate{
>     if (same BA):
>         MultiFab* ptmp = &dst;
>     else:
>         create tmp
>         MultiFab* ptmp = &tmp;
>     CoarsenAndInterpolateLoop
>     if (different BA):
>         copy from tmp to dst
> }
> ```
> 
> Just suggesting, we could discuss it.

Thanks for this suggestion. So far I left the intermediate function `CoarsenAndInterpolateLoop` for the following reason: if we want to remove this intermediate function and write the corresponding code block directly in the highest-level function `CoarsenAndInterpolate`, we would need to use a similar logic as the one outlined in your pseudo-code for the integer variable `dcomp` too, as this is passed once with the value `dcomp` and once with the value `0`, depending on the result of the `if` control statements. I can try this and check whether the resulting code is easier or more difficult to understand than the current code at the end.

File: Source/Utils/Average.H (hunk 9 above)
Comment (MaxThevenet) [reply Source/Utils/Average.H]: No need to put it there again.

File: Source/Utils/Average.cpp [hunk 16]
```diff
@@ -3,27 +3,109 @@
...
+                                     const int scomp,
+                                     const int ncomp,
+                                     const int ngrow,
+                                     const IntVect crse_ratio )
 {
+    // Staggering of source fine MultiFab and destination coarse MultiFab
+    const IntVect stag_src = mf_src.boxArray().ixType().toIntVect();
+    const IntVect stag_dst = mf_dst.boxArray().ixType().toIntVect();
+
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE( mf_src.nGrowVect() >= IntVect(stag_dst-stag_src),
```
Comment (MaxThevenet) [inline Source/Utils/Average.cpp:18]:
```
    IntVect(stag_dst-stag_src),
```
Isn't the difference between two IntVect already an IntVect?

File: Source/Utils/Average.cpp [hunk 17]
```diff
@@ -3,27 +3,109 @@
...
+    int const* const AMREX_RESTRICT sf = sf_gpuarr.data();
+    int const* const AMREX_RESTRICT sc = sc_gpuarr.data();
+    int const* const AMREX_RESTRICT cr = cr_gpuarr.data();
+    int      * const AMREX_RESTRICT np = np_gpuarr.data();
+
+    // Compute number of points to loop over (either 1 or 2) in each direction
+    for ( int l = 0; l < 3; ++l ) {
+        if ( cr[l] == 1 ) np[l] = 1+abs(sf[l]-sc[l]); // no coarsening
+        else              np[l] = 2-sf[l];
+    }
```
Comment (MaxThevenet) [inline Source/Utils/Average.cpp:61]: Doesn't this belong to `CoarsenAndInterpolateKernel` instead?

File: Source/Utils/Average.cpp (hunk 16 above)
Comment (MaxThevenet) [reply Source/Utils/Average.cpp:18]: Actually I think this check should include the number of guard cells required too...
Is there an example where we need to update the guard cells? Otherwise, I am fine with `AMREX_ALWAYS_ASSERT(ngrow==0)`, as it may be difficult to have the right number of guard cells, so maybe we would do differently (update valid cells only, and then exchanged guard cells on the coarsened MultiFab).

File: Source/Utils/Average.cpp (hunk 17 above)
Comment (EZoni) [reply Source/Utils/Average.cpp:61]: It can be in the kernel, but my idea here was that we can compute this number outside the loop over the boxes and outside the `ParallelFor`, because it is a quantity that logically depends only on the staggering of the source and destination `MultiFab`s. There is no significant gain in performance, as we have this loop in the kernel anyways in order to compute the minimum index in each direction. If you think that the previous implementation was clearer, I can move it back to the kernel. I was thinking that one may use the possibility to give the number of points as an input argument to the kernel function also to try to have one single kernel function for both averaging and interpolation in the future (since the two would differ only for the number of points and the weight of each point). But this is also not necessary, so I can go back to the previous implementation here.

File: Source/Utils/Average.cpp (hunk 16 above)
Comment (EZoni) [reply Source/Utils/Average.cpp:18]: Yes, the compiler does not complain if I leave simply `stag_dst-stag_src`, so I can modify the code accordingly. Regarding the guard cells, I have encountered problems when I do not fill the guard cells (in the equivalent previous calls to `Average::ToCellCenter`) in tests using back-transformed diagnostics (not yet sure of the reason, though).

File: Source/Utils/Average.cpp (hunk 17 above)
Comment (MaxThevenet) [reply Source/Utils/Average.cpp:61]: Sounds legit indeed, let's keep it this way.
